// thingandlink/tuyarollertai.js
const tasmotatai = require('./tasmotatai');
const { items } = require('ohj');
const log = require('ohj').log('tuyarollertai');

class TasmotaDimmerTAI extends tasmotatai.AbstractTasmotaTAI {
    buildObjects() {
        super.buildObjects();
        this.addDimmer();
        this.buildSignalTAIs();
    }

    addOpenCloseItems(){

        let openCloseChannel = this.buildOpenCloseChannel();

        let openCloseItem = items.createItem(`${this.id}_Dimmer`, 'Dimmer', 'roller', this.groups, `${this.name} Dimmer`, ["Tasmota"]);
        
        this.linkItemToChannel(openCloseItem, openCloseChannel, {
            autoupdate: 'false'
        });

        this.openCloseItem = openCloseItem; //used for sitemap
    } 

    buildOpenCloseChannel() {

        return this.withNewMQTTChannel(`STATE`, 'Dimmer', 'dimmer', {
            stateTopic: `stat/${this.mqttId}/POSITION`, 
            commandTopic: `cmnd/${this.mqttId}/Dimmer`
        });
    }
}

class TuyaRollerTAI extends TasmotaDimmerTAI {

}

module.exports = {
    TuyaRollerTAI
}
