// mythings/tasmota.js
const { TasmotaTAI, FourChannelTasmotaTAI, TuyaRollerTAI } = require('thingandlink').device;

class HallwayLightsTAI extends TasmotaTAI {
    constructor() {
        super({
            name: 'HallwayLights', 
            mqttId: 'sonoff-2DDA40',
            grousps: ["gHallway"],
            tags: ["Lightbulb"]
        });
    }
    
    buildSwitchChannel() {
        return this.withNewMQTTChannel('POWER', 'Switch', 'switch', {
            stateTopic: `stat/${this.mqttId}/lights`, 
            commandTopic: `cmnd/${this.mqttId}/POWER1`,
            on:"TOGGLE",
            off:"TOGGLE",
            transformationPattern:"MAP:numberToOnOff.map"
        });
    }
}

module.exports = [
    new TasmotaTAI({name: 'Xmas Tree', mqttId: 'sonoff-6E29EB', groups:['gLivingRoom'], tags: ["LightStripe"]}),
    new TasmotaTAI({name: 'GarageCarDoor', mqttId: 'sonoff-095C5F', groups:['gGarageCarDoor']}),
    new TasmotaTAI({name: 'GarageLights', mqttId: 'sonoff-14FA0A', groups:['gGarage'], tags: ["Lightbulb"]}),
    new TasmotaTAI({name: 'OutdoorLights', mqttId: 'sonoff-66A06B', tags: ["Lightbulb"]}),
    new TasmotaTAI({name: 'GarageSideDoor', mqttId: 'sonoff-14FA37', tags: ["SideDoor"]}),
    new TasmotaTAI({name: 'Flamingo', mqttId: 'sonoff-3B47B7', tags: ["LightStripe"], groups:['gLivingRoom']}),
    new TasmotaTAI({name: 'Unused', mqttId: 'sonoff-3B4A56'}),
    new TasmotaTAI({name: 'PoolPump', mqttId: 'sonoff-3ED374', groups:['gGarage'], tags: ["Pump"]}),
    new TasmotaTAI({name: 'UpstairsDoorLatch', mqttId: 'sonoff-56D77C', groups:["gUpstairsBedroom"], tags: ["InnerDoor"]}),

    new TasmotaTAI({name: 'BedPump', mqttId: 'tasmota_440B88', groups:["gUpstairsBedroom"], tags: ["Pump"]}),

    new HallwayLightsTAI(),
    
    new FourChannelTasmotaTAI({name: "ACDucts", mqttId: "sonoff-CAA8AF", groups:["gBluehouse"], tags: ["HVAC"]}),
    new FourChannelTasmotaTAI({name: "OutdoorWater", mqttId: "sonoff-DB52A6", groups:["gOutdoor"], tags: ["Pump"]}),

    //todo: move out
    // new TuyaRollerTAI({name: 'Games Room Door Roller', mqttId: 'tasmota_7DAB5C', groups:["gUpstairs"]}),

];
